crates/wasmparser/src/binary_reader.rs

```diff
@@ -242,78 +300,458 @@ impl<'a> BinaryReader<'a> {
         })
     }
 
+    pub(crate) fn read_type_def(&mut self) -> Result<TypeDef> {
+        Ok(match self.read_u8()? {
+            0x60 => TypeDef::Func(self.read_func_type()?),
+            x => {
+                return Err(BinaryReaderError::new(
+                    format!("invalid leading byte (0x{:x}) for type", x),
+                    self.original_position() - 1,
+                ))
+            }
+        })
+    }
+
+    pub(crate) fn read_component_type_def(&mut self) -> Result<ComponentTypeDef<'a>> {
+        Ok(match self.read_u8()? {
+            0x4f => ComponentTypeDef::Module(
+                (0..self.read_var_u32()?)
+                    .map(|_| self.read_module_type())
+                    .collect::<Result<_>>()?,
+            ),
+            0x4e => ComponentTypeDef::Component(
+                (0..self.read_var_u32()?)
+                    .map(|_| self.read_component_type())
+                    .collect::<Result<_>>()?,
+            ),
+            0x4d => ComponentTypeDef::Instance(
+                (0..self.read_var_u32()?)
+                    .map(|_| self.read_instance_type())
+                    .collect::<Result<_>>()?,
+            ),
+            0x4c => ComponentTypeDef::Function(ComponentFuncType {
+                params: (0..self.read_var_u32()?)
+                    .map(|_| Ok((self.read_string()?, self.read_interface_type()?)))
+                    .collect::<Result<_>>()?,
+                result: self.read_interface_type()?,
+            }),
+            0x4b => ComponentTypeDef::Value(self.read_interface_type()?),
+            // Delegate all other values to `read_compound_type`.
+            x => ComponentTypeDef::Compound(self.read_compound_type(x)?),
+        })
+    }
+
     pub(crate) fn read_module_type(&mut self) -> Result<ModuleType<'a>> {
-        let pos = self.original_position();
-        let imports_len = self.read_var_u32()? as usize;
-        if imports_len > MAX_WASM_IMPORTS {
-            return Err(BinaryReaderError::new("imports size is out of bounds", pos));
-        }
-        Ok(ModuleType {
-            imports: (0..imports_len)
-                .map(|_| self.read_import())
-                .collect::<Result<_>>()?,
-            exports: self.read_export_types()?,
+        Ok(match self.read_u8()? {
+            0x01 => ModuleType::Type(self.read_type_def()?),
+            0x02 => ModuleType::Import(self.read_import()?),
+            0x07 => ModuleType::Export {
+                name: self.read_string()?,
+                ty: self.read_type_ref()?,
+            },
+            x => {
```

Comment:
Aliases aren't supported in module types as outer aliases can only refer to component function types because components have no core function types in their type space (and parent index spaces are always components).

Thus with the current spec, module types are forced to encode their function types "inline".

Reply:
Marking resolved for now (I'll go through every comment after the PR's finished to document spec questions/issues to raise).